test-apps/sediment/src/main/java/com/twosigma/waiter/sediment/Servlet.java

```diff
@@ -0,0 +1,225 @@
+
+        final Map<String, Object> info = new HashMap<>();
+
+        info.put("body-length", String.valueOf(bodyLength));
+        if (request.getContentLengthLong() >= 0) {
+            info.put("content-length", String.valueOf(request.getContentLengthLong()));
+        }
+        info.put("context-path", request.getContextPath());
+        info.put("headers",
+            convertFieldsToMap(
+                Collections.list(request.getHeaderNames()),
+                (name) -> {
+                    final List<String> headerValues = Collections.list(request.getHeaders(name));
+                    if (headerValues.size() == 1) {
+                        return headerValues.get(0);
+                    } else {
+                        return headerValues;
+                    }
+                }));
+        info.put("path-info", request.getPathInfo());
+        info.put("protocol", request.getProtocol());
+        info.put("method", request.getMethod());
+        info.put("query-parameters",
+            convertFieldsToMap(
+                Collections.list(request.getParameterNames()),
+                (name) -> {
+                    final List<String> headerValues = Arrays.asList(request.getParameterValues(name));
+                    if (headerValues.size() == 1) {
+                        return headerValues.get(0);
+                    } else {
+                        return headerValues;
+                    }
+                }
+            ));
+        info.put("request-uri", request.getRequestURI());
+        info.put("scheme", request.getScheme());
+        final HttpFields requestTrailers = request.getTrailers();
+        if (requestTrailers != null) {
+            info.put("trailers",
+                convertFieldsToMap(
+                    requestTrailers.getFieldNamesCollection(),
+                    (name) -> {
+                        final List<String> trailerValues = Collections.list(requestTrailers.getValues(name));
+                        if (trailerValues.size() == 1) {
+                            return trailerValues.get(0);
+                        } else {
+                            return trailerValues;
+                        }
+                    }));
+        }
+        info.put("user-principal", request.getUserPrincipal() != null ? request.getUserPrincipal().toString() : "");
+
+        return info;
+    }
+
+    private static HttpFields createResponseTrailers(final Request request) {
+        final HttpFields trailerFields = new HttpFields();
+        final String trailerPrefix = "x-sediment-response-trailer-";
+        for (String name : Collections.list(request.getHeaderNames())) {
+            if (name.toLowerCase().startsWith(trailerPrefix)) {
+                final String trailerName = name.substring(trailerPrefix.length());
+                trailerFields.add(trailerName, request.getHeader(name));
+            }
+        }
+        if (trailerFields.size() == 0) {
+            return null;
```

Comment:
`null` means we are not going to sending trailers, whereas a non-null return means we will send possibly empty trailers. Is not a big deal either way and would prefer not to change the code.